File: ios/SkinnyLegend/App/AppMode.swift
import Foundation

/// Decides whether the app talks to the real backend or to the in-memory mock. Normally that is
/// settled once per launch by the `-mockAPI` argument; a DEBUG build can also flip into the mock
/// at runtime from the sign-in screen (see `AppModeStore`), which is what `mockOverride` carries.
enum AppMode {
    /// Which set of services `AppEnvironment.make()` should stand up.
    enum Services: Equatable, Sendable {
        case live
        case mock
    }

    /// Where the DEBUG "Dùng dữ liệu mẫu" override is persisted.
    static let mockOverrideKey = "app.mockOverride"

    /// Set by the `SkinnyLegend` scheme's default launch arguments, and by
    /// `xcrun simctl launch booted com.themarcus125.skinnylegend -mockAPI`.
    static var launchArgumentIsMock: Bool {
        ProcessInfo.processInfo.arguments.contains("-mockAPI")
    }

    /// The runtime override, off unless a DEBUG build turned it on.
    ///
    /// Compiled out of Release entirely — both halves. The flag lives in `UserDefaults`, which
    /// survives a Debug-to-Release reinstall on the same device, so a Release build that still
    /// *read* it could boot straight into the mock with no UI anywhere to turn it off (the
    /// "Dùng dữ liệu mẫu" control is itself `#if DEBUG`). With the getter gone, `services(...)`
    /// provably receives `false` for `mockOverride` in Release.
    static var mockOverride: Bool {
        get {
            #if DEBUG
            UserDefaults.standard.bool(forKey: mockOverrideKey)
            #else
            false
            #endif
        }
        set {
            #if DEBUG
            UserDefaults.standard.set(newValue, forKey: mockOverrideKey)
            #endif
        }
    }

    static var isMock: Bool {
        launchArgumentIsMock || mockOverride
    }

    static var hasFirebasePlist: Bool {
        Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil
    }

    /// Firebase is configured, and the live API client used, only here.
    static var useLiveBackend: Bool {
        services(launchArgumentIsMock: launchArgumentIsMock,
                 mockOverride: mockOverride,
                 hasFirebasePlist: hasFirebasePlist) == .live
    }

    /// The whole decision, as one pure function over the three inputs, so it can be unit tested
    /// without touching `ProcessInfo`, `Bundle` or `UserDefaults`. Either mock signal wins, and a
    /// launch with no `GoogleService-Info.plist` must land on the mock rather than stand up
    /// `FirebaseAuthService` against an unconfigured `FirebaseApp` and crash.
    static func services(launchArgumentIsMock: Bool, mockOverride: Bool, hasFirebasePlist: Bool) -> Services {
        servicesAreLive(isMock: launchArgumentIsMock || mockOverride, hasFirebasePlist: hasFirebasePlist)
            ? .live
            : .mock
    }

    /// Pure decision behind `services`, kept as its own step because "is this run mock?" and "is
    /// Firebase configured?" are independent reasons to fall back.
    static func servicesAreLive(isMock: Bool, hasFirebasePlist: Bool) -> Bool {
        !isMock && hasFirebasePlist
    }

    /// Environment first (a scheme, or a device on the same Wi-Fi pointing at a laptop running
    /// the Hono API), then the bundle's `API_BASE_URL` — fed per configuration from
    /// `project.yml`, so a Release/TestFlight build resolves to the deployed API instead of
    /// `localhost`. The literal remains only as a last resort for a bundle without the key.
    static var apiBaseURL: URL {
        resolveBaseURL(
            environment: ProcessInfo.processInfo.environment,
            infoPlist: Bundle.main.infoDictionary ?? [:]
        )
    }

    /// Pure resolution order behind `apiBaseURL`, injectable for tests: a set but unparseable or
    /// empty value at one level falls through to the next rather than crashing the launch.
    static func resolveBaseURL(environment: [String: String], infoPlist: [String: Any]) -> URL {
        if let url = url(from: environment["API_BASE_URL"]) { return url }
        if let url = url(from: infoPlist["API_BASE_URL"] as? String) { return url }
        return URL(string: "http://localhost:3000")!
    }

    private static func url(from raw: String?) -> URL? {
        guard let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty,
              let url = URL(string: trimmed), url.scheme != nil else { return nil }
        return url
    }

    static var appVersion: String {
        let short = Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "1.0"
        let build = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "1"
        return "\(short) (\(build))"
    }

    /// Group fund page (spec v1.1 §A). Per-challenge, so a constant rather than an API field.
    static let momoFundURL = URL(string: "https://quy.momo.vn/v2/GZqk7REIhy?cover=f131")!
}

File: apps/web/src/lib/app-mode.ts
/**
 * Whether this load talks to the real backend or to `@skinny/api-client/mock`.
 *
 * Port of `AppMode` (`ios/SkinnyLegend/App/AppMode.swift`). iOS settles the question with a
 * `-mockAPI` launch argument plus a DEBUG-only `UserDefaults` override flipped from the sign-in
 * screen; the web reads `VITE_MOCK` (inlined at build time) plus the same runtime override, kept
 * in `localStorage`.
 */
export type Services = 'live' | 'mock';

/**
 * Port of `AppMode.services`. Either mock signal wins, and a build with no Firebase config must
 * land on the mock rather than stand up `initializeAuth` against an unconfigured app and throw.
 *
 * Pure over its three inputs so it can be tested without `import.meta.env` or storage — exactly
 * why the Swift original takes its `ProcessInfo`/`Bundle`/`UserDefaults` reads as parameters.
 */
export function resolveServices({
  envIsMock,
  mockOverride,
  hasFirebaseConfig,
}: {
  envIsMock: boolean;
  mockOverride: boolean;
  hasFirebaseConfig: boolean;
}): Services {
  return !(envIsMock || mockOverride) && hasFirebaseConfig ? 'live' : 'mock';
}

/** Where the dev-only "Dùng dữ liệu mẫu" override is persisted. `AppMode.mockOverrideKey`. */
export const MOCK_OVERRIDE_KEY = 'skinny.mockOverride';

function safeStorage(storage?: Storage): Storage | undefined {
  if (storage) return storage;
  // Safari in private mode throws on `localStorage` access rather than returning null.
  try {
    return typeof localStorage === 'undefined' ? undefined : localStorage;
  } catch {
    return undefined;
  }
}

/**
 * The runtime override, off unless a dev build turned it on.
 *
 * Both halves are gated on `import.meta.env.DEV`, mirroring the `#if DEBUG` on *both* the getter
 * and the setter in Swift: the flag lives in `localStorage`, which survives a dev-to-production
 * deploy on the same origin, so a production build that still *read* it could boot straight into
 * the mock with no UI anywhere to turn it off (the "Dùng dữ liệu mẫu" control is itself
 * dev-only). With the read compiled out, `resolveServices` provably gets `false` in production.
 */
export function readMockOverride(storage?: Storage): boolean {
  if (!import.meta.env.DEV) return false;
  try {
    return safeStorage(storage)?.getItem(MOCK_OVERRIDE_KEY) === '1';
  } catch {
    return false;
  }
}

export function writeMockOverride(value: boolean, storage?: Storage): void {
  if (!import.meta.env.DEV) return;
  try {
    const store = safeStorage(storage);
    if (value) store?.setItem(MOCK_OVERRIDE_KEY, '1');
    else store?.removeItem(MOCK_OVERRIDE_KEY);
  } catch {
    // A full or blocked quota must not take the sign-in screen down with it.
  }
}
